### PR TITLE
Enable aurae to operate on TCP or Unix sockets based on context

### PR DESCRIPTION
**Pull request by dmah42**
also removes musl and upgrades cgroups-rs so it builds for me locally.  obviously i expect this to be resolved on main so i can rebase to it.

### CODE REVIEW

File: auraed/src/init/system_runtimes/mod.rs [hunk 1]
```diff
@@ -28,18 +34,82 @@
...
+    ) -> Result<SocketStream, InitError>;
+}
+
+async fn create_unix_socket_stream(
+    socket_path: PathBuf,
+) -> Result<SocketStream, InitError> {
+    let _ = std::fs::remove_file(&socket_path);
+    let sock_path = Path::new(&socket_path)
+        .parent()
+        .ok_or_else(|| anyhow!("unable to find socket path"))?;
```
Comment (future-highway) [inline auraed/src/init/system_runtimes/mod.rs:83]: The wording on this error made me think that the directory needs to exist already. Maybe something like "not a valid socket path" would be better?

File: auraed/src/lib.rs [hunk 2]
```diff
@@ -192,18 +193,15 @@ struct AuraedRuntime {
...
-            )
-        })?;
+    pub async fn run<T, IO, IE>(
+        &self,
+        socket_stream: T,
+    ) -> Result<(), Box<dyn std::error::Error>>
+    where
+        T: tokio_stream::Stream<Item = Result<IO, IE>> + Send + 'static,
+        IO: AsyncRead + AsyncWrite + Connected + Unpin + Send + 'static,
+        IE: Into<Box<dyn std::error::Error + Send + Sync>>,
```
Comment (future-highway) [inline auraed/src/lib.rs:203]: Congratulations :)

**Review by future-highway — COMMENTED** [review]
I assume we don't want to merge this yet, pending musl?

File: aurae-client/src/client.rs [hunk 3]
```diff
@@ -108,13 +108,19 @@ impl AuraeClient {
...
+            Channel::builder(uri).tls_config(tls_config)?.connect().await
+        } else {
+            Channel::from_static(KNOWN_IGNORED_SOCKET_ADDR)
+                .tls_config(tls_config)?
+                .connect_with_connector(service_fn(move |_: Uri| {
+                    UnixStream::connect(system.socket.clone())
+                }))
+                .await
+        }
+        .with_context(|| "unable to connect auraed system socket")?;
```
[the remark this answers is not in the document]
Comment (dmah42) [reply aurae-client/src/client.rs:123]: added a tweak in this iteration.

File: auraed/src/init/mod.rs [hunk 4]
```diff
@@ -70,21 +71,34 @@ pub(crate) enum InitError {
     Fs(#[from] FsError),
     #[error(transparent)]
     Network(#[from] NetworkError),
+    #[error(transparent)]
+    Anyhow(#[from] anyhow::Error),
```
[the remark this answers is not in the document]
Comment (dmah42) [reply auraed/src/init/mod.rs:75]: fixed the errors. i didn't like how i'd added a load of things to the nice neat `InitError` anyway.

File: auraed/src/init/mod.rs [hunk 5]
```diff
@@ -70,21 +71,34 @@ pub(crate) enum InitError {
...
+    #[error(transparent)]
+    IO(#[from] std::io::Error),
```
[the remark this answers is not in the document]
Comment (dmah42) [reply auraed/src/init/mod.rs:77]: ack

File: auraed/src/init/mod.rs [hunk 6]
```diff
@@ -70,21 +71,34 @@ pub(crate) enum InitError {
...
+            ContainerSystemRuntime {}.init(verbose, socket_address)
+        }
+        Context::Daemon => DaemonSystemRuntime {}.init(verbose, socket_address),
     }
     .await;
 
     if let Err(e) = init_result {
         panic!("Failed to initialize: {:?}", e)
     }
+    init_result.expect("this is impossible as we check for error just above")
```
[the remark this answers is not in the document]
Comment (dmah42) [reply auraed/src/init/mod.rs:101]: ack

File: auraed/src/init/system_runtimes/mod.rs [hunk 7]
```diff
@@ -28,18 +34,82 @@
...
+    /// Contains a stream for listening over a TCP socket.
+    Tcp {
+        /// A stream that listens over a TCP socket.
+        stream: TcpListenerStream,
+    },
+
+    /// Contains a stream for listening over a Unix socket.
+    Unix {
+        /// A stream that listens over a Unix socket.
+        stream: UnixListenerStream,
```
[the remark this answers is not in the document]
Comment (dmah42) [reply auraed/src/init/system_runtimes/mod.rs:64]: ack

File: auraed/src/init/system_runtimes/mod.rs (hunk 1 above)
Comment (dmah42) [reply auraed/src/init/system_runtimes/mod.rs:83]: ack

File: auraed/src/lib.rs [hunk 8]
```diff
@@ -124,9 +122,13 @@ struct AuraedOptions {
     /// The CA certificate. Defaults to /etc/aurae/pki/ca.crt
     #[clap(long, value_parser, default_value = "/etc/aurae/pki/ca.crt")]
     ca_crt: String,
-    /// Aurae socket path. Defaults to /var/run/aurae/aurae.sock
-    #[clap(short, long, value_parser, default_value = AURAE_SOCK)]
-    socket: String,
+    /// Aurae socket address.  Depending on context, this should be a file or a network address.
+    /// Defaults to /var/run/aurae/aurae.sock or [::1]:8080 respectively.
```
[the remark this answers is not in the document]
Comment (dmah42) [reply auraed/src/lib.rs:126]: ack

**Review by future-highway — APPROVED** [review]

**Pull request merged**